src/battle/overlay_12_0224E4FC.c

```diff
@@ -5541,3 +5542,842 @@ BOOL TryFling(BattleSystem *bsys, BATTLECONTEXT *ctx, int battlerId) {
+                dmg /= 2;
+                break;
+            case TYPE_WATER:
+                dmg = dmg * 15 / 10;
+                break;
+            }
+        }
+        
+        if ((fieldCondition & FIELD_CONDITION_WEATHER_NO_SUN) && moveNo == MOVE_SOLAR_BEAM) {
+            dmg /= 2;
+        }
+        if (fieldCondition & FIELD_CONDITION_SUN_ALL) {
+            switch(moveType) {
+            case TYPE_FIRE:
+                dmg = dmg * 15 / 10;
+                break;
+            case TYPE_WATER:
+                dmg /= 2;
+                break;
+            }
+        }
+    }
+    
+    if (GetBattlerVar(ctx, battlerIdAttacker, BMON_DATA_FLASH_FIRE_ACTIVE, NULL) && moveType == TYPE_FIRE) {
+        dmg = dmg * 15 / 10;
+    }
+    
+    return dmg + 2;
+}
+
+int ApplyDamageRange(BattleSystem *bsys, BATTLECONTEXT *ctx, int damage) {
+    if (damage) {
+        damage *= (100 - (BattleSystem_Random(bsys) % 16));
+        damage /= 100;
+        if (!damage) {
+            damage = 1;
+        }
+    }
+    return damage;
+}
+
+extern u8 sCritChance[5];
+
+u32 TryCriticalHit(BattleSystem *bsys, BATTLECONTEXT *ctx, int battlerIdAttacker, int battlerIdTarget, int critCnt, u32 sideCondition) {
+    u16 critUp;
+    int item;
+    u16 species;
+    u32 status2;
+    u32 moveEffect;
+    int ret = 1;
+    int ability;
+    
+    item = GetItemVar(ctx, GetBattlerHeldItem(ctx, battlerIdAttacker), ITEM_VAR_HOLD_EFFECT);
+    species = ctx->battleMons[battlerIdAttacker].species;
+    status2 = ctx->battleMons[battlerIdAttacker].status2;
+    moveEffect = ctx->battleMons[battlerIdTarget].moveEffectFlags;
+    ability = ctx->battleMons[battlerIdAttacker].ability;
+    
+    critUp = (((status2 & STATUS2_FOCUS_ENERGY) != 0)*2) + 
+             (item == HOLD_EFFECT_CRITRATE_UP) +
+             critCnt +
+            (ability == ABILITY_SUPER_LUCK) +
+             2*((item == HOLD_EFFECT_CHANSEY_CRITRATE_UP) && (species == SPECIES_CHANSEY)) +
+             2*((item == HOLD_EFFECT_FARFETCHD_CRITRATE_UP) && (species == SPECIES_FARFETCHD));
+             
+    if (critUp > 4) {
+        critUp = 4;
+    }
+    
+    if ((BattleSystem_Random(bsys) % sCritChance[critUp]) == 0) {
+        if (!CheckBattlerAbilityIfNotIgnored(ctx, battlerIdAttacker, battlerIdTarget, ABILITY_BATTLE_ARMOR) && !CheckBattlerAbilityIfNotIgnored(ctx, battlerIdAttacker, battlerIdTarget, ABILITY_SHELL_ARMOR) &&
+            !(sideCondition & SIDE_CONDITION_12) && !(moveEffect & MOVE_EFFECT_LUCKY_CHANT)) {
+            ret = 2;
+        }
+    }
+    
+    if ((ret == 2) && GetBattlerAbility(ctx, battlerIdAttacker) == ABILITY_SNIPER) {
+        ret = 3;
+    }
+    
+    return ret;
+}
+
+extern u16 sMetronomeUnuseableMoves[];
+
+BOOL CheckLegalMimicMove(u16 moveNo) {
+    int i = 0;
+    
+    do {
+        if (sMetronomeUnuseableMoves[i] == moveNo) {
+            break;
+        }
+        i++;
+    } while (sMetronomeUnuseableMoves[i] != 0xFFFE);
```

Comment:
can this not be a for statement?

Reply:
I remember having trouble matching it like that but I can try again